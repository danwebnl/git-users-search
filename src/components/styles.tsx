import styled from 'styled-components';

export const MainContainer = styled.div`
    margin: 0px auto;
    margin-bottom: 20px;
    width: 600px;
    font-family: Arial, Helvetica, sans-serif;
    color: #4a494a;
    border: 1px solid #a5a5a5;
    text-align: center;

    h1 {
        border-bottom: 1px solid #a5a5a5;
        padding: 20px;
        margin: 0px;
    }
`;

export const Form = styled.form`

    border-bottom: 0px;

    label {
        font-weight: bold;
        font-size: 16px;
        padding: 20px;
    }

    select {
        width: 100%;
        font-size: 15px;
        margin-top: 10px;
    }

    input[type='text'] {
        width: 90%;
        height: 30px;
        padding: 5px;
        margin: 0px auto;
        font-size: 15px;
        margin-top: 10px;
        margin-bottom: 10px;
    }
`;

export const FormSubmitContainer = styled.div`
    padding: 20px;
    border-top: 1px solid #a5a5a5;

    button {
        width: 100%;
        margin-top: 10px;
        background-color: #eee;
        border: 1px solid #a5a5a5;
        font-weight: bold;
        font-size: 15px;
        color: #4a494a;
        padding: 10px;
        border-radius: 5px;
        cursor: pointer;
        &:disabled {
            color: #cecece;
            background: #fff;
            border: 1px solid #cecece;
        }
    }
`;

export const Error = styled.div`
    color: red;
    padding: 20px;
`;

export const ListContainer = styled.div`
    padding: 20px;
    border-top: 1px solid #a5a5a5;

    ul {
        list-style-type: none;
        padding: 0px;
    }
`;

export const Loading = styled.div`
    font-style: italic;
    padding: 20px;
    font-size: 13px;
    border-top: 1px solid #a5a5a5;
`;

export const BackButtonContainer = styled.div`
    padding: 20px;
    text-align: left;
    font-size: 13px;
    border-bottom: 0px solid #a5a5a5;
`;
