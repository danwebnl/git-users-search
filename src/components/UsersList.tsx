import * as React from 'react';
import { Link } from 'react-router-dom';

import { ListContainer } from './styles';

interface IProps {
    users: [{
        login: string
    }];
}

const UsersList = (props: IProps) => (
    <ListContainer>
        <h2>Search Results</h2>
        <ul>
            {
                props.users.map((user: { login: string }, index: number) => (
                    <li key={`user${index}`}>
                        <Link to={`/user/${user.login}`}>{user.login}</Link>
                    </li>
                ))
            }
        </ul>
    </ListContainer>
);

export default UsersList;
