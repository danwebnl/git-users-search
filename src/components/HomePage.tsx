import * as React from 'react';
import { connect } from 'react-redux';

import UsersList from './UsersList';
import { addUsers, reset, displayLoader } from '../store/actions/index';
import { MainContainer, Form, FormSubmitContainer, Error, Loading } from './styles';

interface IProps {
    onAddUsers: (searchValue: string) => void;
    onReset: () => void;
    onDisplayLoader: () => void;
    users: [{
        login: string
    }];
    totalCount: number;
    errorMessage: string;
    displayedLoader: boolean;
}

class HomePage extends React.Component<IProps, { searchValue: string }> {

    constructor(props: any) {
        super(props);

        this.state = {
            searchValue: ''
        };
    }

    handleNameChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
        this.setState({ searchValue: event.target.value });
    }

    handleSubmit = (event: React.ChangeEvent<HTMLFormElement>): void => {
        event.preventDefault();

        if (this.state.searchValue.trim() === '') {
            return;
        }

        const { onAddUsers, onDisplayLoader, onReset } = this.props;
        onReset();
        onDisplayLoader();
        onAddUsers(this.state.searchValue);
    }

    handleReset = (): void => {

        this.setState({
            searchValue: ''
        });

        const { onReset } = this.props;
        onReset();
    }

    render() {
        const { users, totalCount, errorMessage, displayedLoader } = this.props;

        return (
            <MainContainer>
                <h1>Git Users Search</h1>

                <Form onSubmit={this.handleSubmit} method="POST">
                    <input
                        type="text"
                        name="name"
                        placeholder="Insert name here..."
                        value={this.state.searchValue}
                        onChange={this.handleNameChange}
                    />
                    <FormSubmitContainer>
                        <button type="submit" id="submitButton">Search</button>
                        <button type="button" onClick={this.handleReset}>Reset</button>
                    </FormSubmitContainer>
                </Form>

                {errorMessage && (
                    <Error>{errorMessage}</Error>
                )}

                {displayedLoader && (
                    <Loading>Loading...</Loading>
                )}

                {totalCount === 0 && (
                    <h2>No results.</h2>
                )}

                {totalCount > 0 && (
                    <UsersList users={users} />
                )}
            </MainContainer >
        );
    }

}

const mapStateToProps = (state: any) => {
    return {
        users: state.users.users,
        selectedUser: state.users.selectedUser,
        totalCount: state.users.totalCount,
        errorMessage: state.users.errorMessage,
        displayedLoader: state.users.displayedLoader
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return {
        onAddUsers: (searchValue: string) => dispatch(addUsers(searchValue)),
        onReset: () => dispatch(reset()),
        onDisplayLoader: () => dispatch(displayLoader())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
