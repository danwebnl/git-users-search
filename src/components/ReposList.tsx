import * as React from 'react';

import { ListContainer } from './styles';
import { IRepo } from './UserPage';

interface IProps {
    repos: [IRepo];
}

const UsersList = ({ repos }: IProps) => (
    <ListContainer>

        {repos.length < 1 && (
            <h2>No Repositories Found</h2>
        )}

        {repos.length > 0 && (
            <>
                <h2>Repositories </h2>
                <ul>
                    {
                        repos.map((repo: IRepo, index: number) => (
                            <li key={`repo${index}`}>
                                <a href={repo.html_url} target="_blank">{repo.name}</a>
                            </li>
                        ))
                    }
                </ul>
            </>
        )}

    </ListContainer>
);

export default UsersList;
