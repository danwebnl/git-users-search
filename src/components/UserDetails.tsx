import * as React from 'react';

import { ListContainer } from './styles';
import { IUser } from './UserPage';

interface IProps {
    user: IUser;
}

const UserDetails = ({ user }: IProps) => (
    <ListContainer>
        <ul>
            <li>
                <img src={user.avatar_url} title={user.login} style={{ width: 100 }} /></li>
            <li>
                <a href={user.html_url} target="_blank">GitHub Page</a>
            </li>
            <li>Type: {user.type}</li>
            <li>Score: {user.score}</li>
        </ul>

    </ListContainer>
);

export default UserDetails;
