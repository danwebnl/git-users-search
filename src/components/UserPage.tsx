import * as React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { MainContainer, ListContainer, BackButtonContainer, Error } from './styles';
import { getRepos } from '../store/actions/index';
import UserDetails from './UserDetails';
import ReposList from './ReposList';

export interface IUser {
    avatar_url: string;
    html_url: string;
    type: string;
    score: string;
    login: string;
}

export interface IRepo {
    name: string;
    html_url: string;
}

interface IProps {
    user: IUser;
    userLogin: string;
    repos: [IRepo];
    errorMessageRepos: string;
}

class UserPage extends React.Component<IProps, {}> {

    constructor(props: any) {
        super(props);

        const { onGetRepos, user } = props;
        onGetRepos(user.repos_url);
    }

    render() {
        const { user, userLogin, repos, errorMessageRepos } = this.props;

        return (
            <MainContainer>
                <h1>Acount: {userLogin}</h1>

                <BackButtonContainer>
                    <Link to="/">&laquo; Back</Link>
                </BackButtonContainer>

                <UserDetails user={user} />

                {errorMessageRepos && (
                    <Error>{errorMessageRepos}</Error>
                )}

                {repos && (
                    <ReposList repos={repos} />
                )}

            </MainContainer>
        );
    }
}

const mapStateToProps = (state: any, ownProps: any) => {
    return {
        user: state.users.users.find((user: any) => user.login === ownProps.match.params.login),
        userLogin: ownProps.match.params.login,
        repos: state.users.repos,
        errorMessageRepos: state.users.errorMessageRepos
    };
};

const mapDispatchToProps = (dispatch: any) => {
    return {
        onGetRepos: (url: string) => dispatch(getRepos(url))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(UserPage);
