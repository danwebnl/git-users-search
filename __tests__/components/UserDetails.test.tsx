import * as React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import UserDetails from '../../src/components/UserDetails';
import { IUser } from '../../src/components/UserPage';

const testUser: IUser = {
    avatar_url: 'https://avatars0.githubusercontent.com/u/6805019?v=4',
    html_url: 'https://github.com/danwebnl',
    type: 'User',
    score: '1111',
    login: 'danwebnl'
};

describe('<UserDetails />', () => {

    it('renders and matches the snapshot', () => {
        const wrapper = shallow(<UserDetails user={testUser} />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders the image properly', () => {
        const wrapper = shallow(<UserDetails user={testUser} />);
        const img = wrapper.find('img');
        expect(img.props().src).toBe(testUser.avatar_url);
        expect(img.props().title).toBe(testUser.login);
    });

    it('renders the GitHub page URL', () => {
        const wrapper = shallow(<UserDetails user={testUser} />);
        const ATag = wrapper.find('a');
        expect(ATag.children().text()).toBe('GitHub Page');
        expect(ATag.props().href).toBe(testUser.html_url);
    });
});
