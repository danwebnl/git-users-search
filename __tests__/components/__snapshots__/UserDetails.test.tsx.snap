// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDetails /> renders and matches the snapshot 1`] = `
<styled.div>
  <ul>
    <li>
      <img
        src="https://avatars0.githubusercontent.com/u/6805019?v=4"
        style={
          Object {
            "width": 100,
          }
        }
        title="danwebnl"
      />
    </li>
    <li>
      <a
        href="https://github.com/danwebnl"
        target="_blank"
      >
        GitHub Page
      </a>
    </li>
    <li>
      Type: 
      User
    </li>
    <li>
      Score: 
      1111
    </li>
  </ul>
</styled.div>
`;
